src/Po2JsonCommand.php

```diff
@@ -0,0 +1,252 @@
+
+					if ( substr( $file, - 7 ) === '.min.js' ) {
+						return substr( $file, 0, - 7 ) . '.js';
+					}
+
+					if ( substr( $file, - 3 ) === '.js' ) {
+						return $file;
+					}
+
+					return null;
+				},
+				$translation->getReferences()
+			);
+
+			$sources = array_unique( array_filter( $sources ) );
+
+			foreach ( $sources as $source ) {
+				if ( ! isset( $mapping[ $source ] ) ) {
+					$mapping[ $source ] = new Translations();
+					// See https://core.trac.wordpress.org/ticket/45441
+					//$mapping[ $source ]->setDomain( $translations->getDomain() );
+					$mapping[ $source ]->setLanguage( $translations->getLanguage() );
+					$mapping[ $source ]->setHeader( 'PO-Revision-Date', $translations->getHeader( 'PO-Revision-Date' ) );
+
+					$plural_forms = $translations->getPluralForms();
+
+					if ( $plural_forms ) {
+						$mapping[ $source ]->setPluralForms( $plural_forms[0], $plural_forms[1] );
+					}
+				}
+
+				$mapping[ $source ][] = $translation;
+			}
+		}
+
+		$result += $this->build_json_files( $mapping, $base_file_name, $destination );
+
+		return $result;
+	}
+
+	/**
+	 * Builds a mapping of JS file names to translation entries.
+	 *
+	 * Exports translations for each JS file to a separate translation file.
+	 *
+	 * @param array  $mapping        A mapping of files to translation entries.
+	 * @param string $base_file_name Base file name for JSON files.
+	 * @param string $destination    Path to the destination directory.
+	 *
+	 * @return array List of created JSON files.
+	 */
+	protected function build_json_files( $mapping, $base_file_name, $destination ) {
+		$result = [];
+
+		foreach ( $mapping as $file => $translations ) {
+			/** @var Translations $contents */
+
+			$hash             = md5( $file );
+			$destination_file = "${destination}/{$base_file_name}-{$hash}.json";
+
+			$success = JedGenerator::toFile( $translations, $destination_file, [ 'json' => $this->json_options ] );
+
+			if ( ! $success ) {
+				WP_CLI::warning( sprintf( 'Could not create file %s', basename( $destination_file, '.json' ) ) );
+
+				continue;
+			}
+
+			$result[] = $destination_file;
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Removes strings from PO file that only occur in JavaScript file.
+	 *
+	 * @param string $source_file Path to the PO file.
+	 * @return bool True on success, false otherwise.
+	 */
+	protected function remove_js_strings_from_po_file( $source_file ) {
+		/** @var Translations[] $mapping */
+		$translations = new Translations();
+
+		Po::fromFile( $source_file, $translations );
+
+		$translations_to_remove = [];
+
+		foreach ( $translations as $translation ) {
+			/** @var Translation $translation */
+
+			if ( ! $translation->hasReferences() ) {
+				continue;
+			}
+
+			foreach ( $translation->getReferences() as $reference ) {
+				$file = $reference[0];
+
+				if ( substr( $file, - 4 ) === '.php' ) {
+					continue 2;
+				}
+			}
+
+			$translations_to_remove[] = $translation->getId();
```

Comment:
Because the tests are failing when doing that.

I tried both `unset( $translations[ $key ] );` and `unset( $translations[ $translation->getId() ] );`, but in both cases entries are skipped when doing that. Probably has to do with the way `Translations` implements `ArrayObject`.

Reply:
This happens because you modify the array you are iterating over in your `foreach`. As soon as you remove the first item, you mess up the internal iteration pointer.

To solve this, just turn `foreach( $translation as ... )` into `foreach( $translations->getArrayCopy() as ... )`.

Reply:
I just added a commit to show what I mean, tests should still pass now.